webrender/src/device/gfx.rs

```diff
@@ -708,75 +700,68 @@ pub struct Buffer<B: hal::Backend> {
 impl<B: hal::Backend> Buffer<B> {
     pub fn new(
         device: &B::Device,
-        memory_types: &[hal::MemoryType],
-        usage: hal::buffer::Usage,
-        pitch_alignment_mask: usize,
+        heaps: &mut Heaps<B>,
+        memory_usage: MemoryUsageValue,
+        buffer_usage: hal::buffer::Usage,
+        alignment_mask: usize,
         data_len: usize,
         stride: usize,
     ) -> Self {
-        let buffer_size = (data_len * stride + pitch_alignment_mask) & !pitch_alignment_mask;
+        let buffer_size = (data_len * stride + alignment_mask) & !alignment_mask;
         let mut buffer = unsafe {
             device
-                .create_buffer(buffer_size as u64, usage)
+                .create_buffer(buffer_size as u64, buffer_usage)
                 .expect("create_buffer failed")
         };
         let requirements = unsafe { device.get_buffer_requirements(&buffer) };
-        let memory_type = memory_types
-            .iter()
-            .enumerate()
-            .position(|(id, mem_type)| {
-                requirements.type_mask & (1 << id) != 0
-                    && mem_type
-                        .properties
-                        .contains(hal::memory::Properties::CPU_VISIBLE)
-            })
-            .unwrap()
-            .into();
-        let memory = unsafe { device.allocate_memory(memory_type, requirements.size) }
-            .expect("Allocate memory failed");
-        unsafe { device.bind_buffer_memory(&memory, 0, &mut buffer) }
+
+        let alignment = ((requirements.alignment -1) | (alignment_mask as u64)) + 1;
+        let memory_block = heaps.allocate(
+            device,
+            requirements.type_mask as u32,
+            memory_usage,
+            requirements.size,
+            alignment,
+        ).expect("Allocate memory failed");
+
+        unsafe { device.bind_buffer_memory(&memory_block.memory(), memory_block.range().start, &mut buffer) }
             .expect("Bind buffer memory failed");
 
-        let buffer = Buffer {
-            memory,
+        Buffer {
+            memory_block,
             buffer,
             buffer_size: requirements.size as _,
             buffer_len: data_len,
             stride,
             state: Cell::new(hal::buffer::Access::empty()),
-        };
-        buffer
+        }
     }
 
-    pub fn update_all<T: Copy>(&self, device: &B::Device, data: &[T]) {
-        let mut upload_data = unsafe {
-            device.acquire_mapping_writer::<T>(&self.memory, 0 .. self.buffer_size as u64)
+    pub fn update_all<T: Copy>(&mut self, device: &B::Device, data: &[T]) {
+        let size = (data.len() * std::mem::size_of::<T>()) as u64;
+        let range = 0 ..  size;
+        unsafe {
+            let mut mapped = self.memory_block.map(device, range.clone()).expect("Mapping memory block failed");
```

Comment:
does the range here obey `nonCoherentAtomSize` limit?

Reply:
The end probably not, will fix that.